Upstream-Status: Submitted
Signed-off-by: Ross Burton <ross.burton@intel.com>

From 57572c82c1b8a719beb1a702c532d4e71aa9609f Mon Sep 17 00:00:00 2001
From: Ross Burton <ross.burton@intel.com>
Date: Mon, 18 Jun 2018 13:47:16 +0100
Subject: [PATCH] environment: validate cpu_family in cross file

---
 mesonbuild/environment.py | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/mesonbuild/environment.py b/mesonbuild/environment.py
index 5a5c053c..407c73d6 100644
--- a/mesonbuild/environment.py
+++ b/mesonbuild/environment.py
@@ -72,6 +72,22 @@ from .compilers import (
 
 build_filename = 'meson.build'
 
+known_cpu_families = (
+    'aarch64',
+    'arm',
+    'e2k',
+    'ia64',
+    'mips',
+    'mips64',
+    'parisc',
+    'ppc',
+    'ppc64',
+    'ppc64le',
+    'sparc64',
+    'x86',
+    'x86_64'
+)
+
 def detect_gcovr(version='3.1', log=False):
     gcovr_exe = 'gcovr'
     try:
@@ -953,6 +969,10 @@ class CrossBuildInfo:
                     res = eval(value, {'__builtins__': None}, {'true': True, 'false': False})
                 except Exception:
                     raise EnvironmentException('Malformed value in cross file variable %s.' % entry)
+
+                if entry == 'cpu_family' and res not in known_cpu_families:
+                    raise EnvironmentException('Unknown CPU family %s, please report this at https://github.com/mesonbuild/meson/issues/new' % value)
+
                 if self.ok_type(res):
                     self.config[s][entry] = res
                 elif isinstance(res, list):
-- 
2.11.0

