From 126a2932508463d5773b7ed8629bb55114f36ece Mon Sep 17 00:00:00 2001
From: Ross Burton <ross.burton@arm.com>
Date: Tue, 13 Oct 2020 17:38:42 +0100
Subject: [PATCH 6/8] autoreconf-gnuconfigize.patch

---
 bin/autoreconf.in | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/bin/autoreconf.in b/bin/autoreconf.in
index 9ec98650..6ca676c8 100644
--- a/bin/autoreconf.in
+++ b/bin/autoreconf.in
@@ -122,6 +122,7 @@ my $intltoolize = $ENV{'INTLTOOLIZE'} || 'intltoolize';
 my $gtkdocize   = $ENV{'GTKDOCIZE'}   || 'gtkdocize';
 my $autopoint   = $ENV{'AUTOPOINT'}   || 'autopoint';
 my $make        = $ENV{'MAKE'}        || 'make';
+my $gnuconfigize = $ENV{'GNUCONFIGIZE'}       || 'gnu-configize';
 
 # --install -- as --add-missing in other tools.
 my $install = 0;
@@ -639,6 +640,10 @@ sub autoreconf_current_directory ($)
       }
     }
 
+  @ex = grep (/^gnu-configize$/, @exclude);
+  if ($#ex == -1) {
+    xsystem ($gnuconfigize);
+  }
 
   # -------------- #
   # Running make.  #
-- 
2.28.0

