From 68ad2ed0ed004f8410b1590f1d0892223e447932 Mon Sep 17 00:00:00 2001
From: Ross Burton <ross.burton@arm.com>
Date: Tue, 13 Oct 2020 17:38:42 +0100
Subject: [PATCH 08/10] autoreconf-gnuconfigize.patch

---
 bin/autoreconf.in | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/bin/autoreconf.in b/bin/autoreconf.in
index f2d0b3c5..9f4ab610 100644
--- a/bin/autoreconf.in
+++ b/bin/autoreconf.in
@@ -127,6 +127,7 @@ my $intltoolize = $ENV{'INTLTOOLIZE'} || 'intltoolize';
 my $gtkdocize   = $ENV{'GTKDOCIZE'}   || 'gtkdocize';
 my $autopoint   = $ENV{'AUTOPOINT'}   || 'autopoint';
 my $make        = $ENV{'MAKE'}        || 'make';
+my $gnuconfigize = $ENV{'GNUCONFIGIZE'}       || 'gnu-configize';
 
 # --install -- as --add-missing in other tools.
 my $install = 0;
@@ -798,6 +799,11 @@ sub autoreconf_current_directory ($)
       }
     }
 
+  @ex = grep (/^gnu-configize$/, @exclude);
+  if ($#ex == -1) {
+    xsystem ($gnuconfigize);
+  }
+
   # ---------------------------------------------------- #
   # Installing aux files and checking for missing ones.  #
   # ---------------------------------------------------- #
-- 
2.25.1

